File: python/2493. Divide Nodes Into the Maximum Number of Groups.py
class Solution:
    def magnificentSets(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for a, b in edges:
            graph[a-1].append(b-1)
            graph[b-1].append(a-1)

        groups = []
        colour = [0] * n
        def is_bipartite(i, c, group):
            if colour[i]: return colour[i] == c
            colour[i] = c
            group.append(i)
            return all(is_bipartite(neighbour, 3-c, group) for neighbour in graph[i])

        for i in range(n):
            if not colour[i]:
                group = []
                if not is_bipartite(i, 1, group): return -1
                groups.append(group)

        def bfs(node, visited):
            level = 0
            q = deque([node])
            while q:
                size = len(q)
                while size > 0:
                    size -= 1
                    current = q.popleft()
                    if visited[current]: continue
                    visited[current] = True
                    for neighbour in graph[current]:
                        if visited[neighbour]: continue
                        q.append(neighbour)
                level += 1
            return level

        return sum(max(bfs(i, [False]*n) for i in group) for group in groups)

File: java/2493. Divide Nodes Into the Maximum Number of Groups.java
class Solution {
    public int magnificentSets(int n, int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0]-1, k -> new ArrayList<>()).add(edge[1]-1);
            graph.computeIfAbsent(edge[1]-1, k -> new ArrayList<>()).add(edge[0]-1);
        }
        int[] colours = new int[n];
        List<List<Integer>> groups = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (colours[i] == 0) {
                List<Integer> group = new ArrayList<>();
                if (!isBipartite(graph, colours, i, 1, group)) return -1;
                groups.add(group);
            }
        }
        int answer = 0;
        for (List<Integer> group : groups) {
            int lsp = 0;
            for (int i : group) {
                lsp = Math.max(lsp, bfs(graph, i, new boolean[n]));
            }
            answer += lsp;
        }
        return answer;
    }
    private int bfs(Map<Integer, List<Integer>> graph, int node, boolean[] visited) {
        int level = 0;
        Deque<Integer> q = new LinkedList<>();
        q.addLast(node);
        while (!q.isEmpty()) {
            int size = q.size();
            while (size-- > 0) {
                int current = q.pollFirst();
                if (visited[current]) continue;
                visited[current] = true;
                if (!graph.containsKey(current)) continue;
                for (int next : graph.get(current)) {
                    if (visited[next]) continue;
                    q.addLast(next);
                }
            }
            level++;
        }
        return level;
    }
    private boolean isBipartite(
        Map<Integer, List<Integer>> graph,
        int[] colours,
        int i,
        int colour,
        List<Integer> group
    ) {
        if (colours[i] != 0) return colours[i] == colour;
        colours[i] = colour;
        group.add(i);
        if (!graph.containsKey(i)) return true;
        for (int neighbour : graph.get(i)) {
            if (!isBipartite(graph, colours, neighbour, 3-colour, group)) {
                return false;
            }
        }
        return true;
    }
}